web/src/lib/components/shared-components/context-menu/button-context-menu.svelte

```diff
@@ -0,0 +1,124 @@
+  export { getMenuContext };
+</script>
+
+<script lang="ts">
+  import CircleIconButton, { type Color } from '$lib/components/elements/buttons/circle-icon-button.svelte';
+  import ContextMenu from '$lib/components/shared-components/context-menu/context-menu.svelte';
+  import { getContextMenuPosition, type Align } from '$lib/utils/context-menu';
+  import { shortcuts } from '$lib/actions/shortcut';
+  import { focusOutside } from '$lib/actions/focus-outside';
+  import { listNavigation } from '$lib/actions/list-navigation';
+  import { generateId } from '$lib/utils/generate-id';
+  import Portal from '$lib/components/shared-components/portal/portal.svelte';
+
+  export let icon: string;
+  export let title: string;
+  export let align: Align = 'top-left';
+  export let direction: 'left' | 'right' = 'right';
+  export let buttonColor: Color = 'transparent';
+  export let buttonSize: string | undefined = undefined;
+  export let buttonPadding: string | undefined = undefined;
+  export let buttonClass: string | undefined = undefined;
+  export let usePortal: boolean = false;
+
+  let showContextMenu = false;
+  let contextMenuPosition = { x: 0, y: 0 };
+  let menuContainer: HTMLUListElement;
+  let selectedId: string | undefined = undefined;
+  let id = generateId();
+
+  $: buttonId = `context-menu-button-${id}`;
+  $: menuId = `context-menu-${id}`;
+
+  const openDropdown = (event: KeyboardEvent) => {
+    contextMenuPosition = getContextMenuPosition(event, align);
+    showContextMenu = true;
+  };
+
+  const handleClick = (event: MouseEvent) => {
+    contextMenuPosition = getContextMenuPosition(event, align);
+    showContextMenu = !showContextMenu;
+  };
+
+  const handleEnter = (event: KeyboardEvent) => {
+    if (selectedId) {
+      event.preventDefault();
+      const node: HTMLLIElement | null = menuContainer.querySelector(`#${selectedId}`);
+      node?.click();
+      closeDropdown();
+    }
+  };
+
+  const closeDropdown = () => {
+    selectedId = undefined;
+    showContextMenu = false;
+  };
+
+  setContext(() => (showContextMenu = false));
+</script>
+
+<div use:clickOutside={{ onOutclick: closeDropdown }} use:focusOutside={{ onFocusOut: closeDropdown }}>
+  <div
+    use:shortcuts={[
+      {
+        shortcut: { key: 'Enter' },
+        onShortcut: handleEnter,
+        preventDefault: false,
+      },
+    ]}
+    use:listNavigation={{
+      container: menuContainer,
+      selectedId,
+      selectedClass: '!bg-gray-200',
+      openDropdown,
+      closeDropdown,
+      selectionChanged: (node) => (selectedId = node?.id),
+    }}
+  >
+    <CircleIconButton
+      {icon}
+      {title}
+      ariaControls={menuId}
+      ariaExpanded={showContextMenu}
+      ariaHasPopup={true}
+      class={buttonClass}
+      color={buttonColor}
+      id={buttonId}
+      on:click={handleClick}
+      padding={buttonPadding}
+      size={buttonSize}
+    />
+  </div>
+  {#if usePortal}
```

Comment:
I'm wondering if there's a better way to do this without duplicating code? This seems like a use case for Svelte 5 snippets.